src/scene/batching.js

```diff
@@ -0,0 +1,719 @@
+        var node;
+        var ents = [];
+        for(var i=0; i<meshInstances.length; i++) {
+            node = meshInstances[i].node;
+            while(!node._app && node._parent) {
+                node = node._parent;
+            }
+            if (!node._app) continue;
+            // node is entity
+            ents.push(node);
+        }
+        this.register(batch, ents);
+    };
+
+    Batching.prototype.generateBatchesForModels = function(nodes) {
+        var i;
+        var groupMeshInstances = {};
+
+        if (!nodes) {
+            // Full scene
+
+            // delete old batches
+            for(i=0; i<this._batchList.length; i++) {
+                this._batchList[i].refCounter = 1;
+                this.destroyBatch(this._batchList[i]);
+            }
+            this._batchList.length = 0;
+
+            // collect
+            this._collectAndRemoveModels(this.rootNode, groupMeshInstances);
+        } else {
+            // Selected entities
+
+            // delete old batches?
+            // TODO?
+
+            // collect
+            for(i=0; i<nodes.length; i++) {
+                this._collectAndRemoveModels(nodes[i], groupMeshInstances);
+            }
+        }
+
+        var group, lists, groupData, batch;
+        for(var groupId in groupMeshInstances) {
+            if (!groupMeshInstances.hasOwnProperty(groupId)) continue;
+            group = groupMeshInstances[groupId];
+
+            groupData = this._batchGroups[groupId];
+            if (!groupData) {
+                // #ifdef DEBUG
+                console.error("batch group " + groupId + " not found");
+                // #endif
+                continue;
+            }
+
+            lists = this.prepare(group, groupData.isDynamic, groupData.maxAabbSize);
+            for(i=0; i<lists.length; i++) {
+                batch = this.create(lists[i], groupData.isDynamic);
+                this.scene.addModel(batch.model);
+                this._registerEntities(batch, lists[i]);
+            }
+        }
+    };
+
+    /**
+     * @function
+     * @name pc.Batching#prepare
+     * @description Takes a list of mesh instances to be batched and sorts them into lists one for each draw call.
+     */
+    Batching.prototype.prepare = function(meshInstances, isDynamic, maxAabbSize) {
+        if (meshInstances.length === 0) return [];
+        if (maxAabbSize === undefined) maxAabbSize = Number.POSITIVE_INFINITY;
+        var halfMaxAabbSize = maxAabbSize * 0.5;
+        var maxInstanceCount = this.device.supportsBoneTextures ? 1024 : this.device.boneLimit;
+        
+        var i;
+        var material, layer, vertCount, params, params2, param, paramFailed, lightList;
+        var aabb = new pc.BoundingBox();
+        var testAabb = new pc.BoundingBox();
+        
+        var lists = [];
+        var j = 0;
+        var meshInstancesLeft = meshInstances;
+        var meshInstancesLeft2;
```

Comment:
I hate this coding style. Is there a better name for this array?

Reply:
They're getting swapped in the end of the loop. Can call them e.g. A and B.